src/experimental/information_estimates.cpp

```diff
@@ -0,0 +1,401 @@
+                    calculateScanMassProbabilityBetween(0.0f, distances[0])
+                };
+
+                // Appending new measurement outcomes for the current cell
+                appendCellProbabilities(probabilities, cell);
+
+                // Get all the measurement outcomes for the current cell
+                std::vector<std::vector<kt_double>> cell_prob = retrieveCellProbabilities(cell);
+
+                // Compute all the possible combinations for the current cell - algorithm 1
+                std::unordered_map<map_tuple, kt_double, utils::tuple_hash::HashTuple> meas_out_prob = computeMeasurementOutcomesHistogram(cell_prob);
+
+                kt_double cell_mutual_inf = 0.0f;
+                for (auto& pair : meas_out_prob)
+                {
+                    cell_mutual_inf +=  pair.second * measurementOutcomeEntropy(pair.first);
+                }
+
+                // Mutual information of cell x, y given a set of measurements
+                updateCellMutualInformation(1.0 - cell_mutual_inf, cell);
+            }
+        }
+        ++curr_idx;
+    }
+    return m_mutual_grid.sum();
+}
+
+
+void InformationEstimates::appendCellProbabilities(std::vector<kt_double>& measurements, karto::Vector2<int> const & cell)
+{
+    /**
+     * Append measured porbabilities for the given cell
+     * Arguments:
+        * measurements [std::vector<kt_double>]: Vector of LocalizedRangeScan pointers
+        * cell [karto::Vector2<int>]: Cell for appending the data
+     * Return:
+        * Void
+    */
+    std::map<std::vector<int>, std::vector<std::vector<kt_double>>>::iterator it_cell;
+
+    it_cell = m_cell_probabilities.find({cell.GetX(), cell.GetY()});
+    if (it_cell == m_cell_probabilities.end())
+    {
+        // Cell is not present in the map, so append it
+        m_cell_probabilities.insert(std::pair<std::vector<int>, std::vector<std::vector<kt_double>>>(
+            {cell.GetX(), cell.GetY()}, {{measurements[0], measurements[1], measurements[2]}}));
+        m_visited_grid(cell.GetX(), cell.GetY()) = 1;
+    }
+    else
+    {
+        if(m_visited_grid(cell.GetX(), cell.GetY()) == 1)
+        {
+            // Compare the unknown probability, the smallest it is the most information we will have
+            // from the occupied or free state
+            int idx = it_cell->second.size() - 1;
+            if(measurements[2] < it_cell->second[idx][2])
+            {
+                // Replacing
+                it_cell->second[idx][0] = measurements[0];
+                it_cell->second[idx][1] = measurements[1];
+                it_cell->second[idx][2] = measurements[2];
+            }
+        }
+        else
+        {
+            // Cell is already in the map, only add the next measurement outcome
+            it_cell->second.push_back({measurements[0], measurements[1], measurements[2]});
+            m_visited_grid(cell.GetX(), cell.GetY()) = 1;
+        }
```

Comment:
@kmilo7204 hmm, this is dead code, isn't it? `m_visited_grid(cell.GetX(), cell.GetY()) != 1` and `it_cell == m_cell_probabilities.end()` can't be true simultaneously.

Reply:
I had to find words to explain what is going on here. At first it seems that we are not using this code, but in fact this is critical for avoiding appending multiple lectures of the same beam. Please find the explanation below.

This is what is going on:
- In mutualInformationFromScans is a line for cleaning the visited cells m_visited_grid.setZero(), it takes place
            right before we start processing the visited cells for a new scan.
- Some operations are done to calculate the probability of see a cell as occupied, free or unknown.
- The probabilities are given to appendCellProbabilities, so we store them into our map.
- Inside appendCellProbabilities we evaluate if the each cell in the current laser reading has been visited:

    - Since we consider that a cell can be seen only once for one beam of the laser scan. In theory two beams should not see the same cell, but as this is a grid map, it might happen due to the resolution. For avoiding it, we are performing the following evaluation on each of the laser beams of the current laser scan.
        - Has this cell been seen by a beam?. If not then add it to our map, and mark this cells as a visited. EOF the function. **Case 1**
        - Is the cell in the map and has been marked as visited by a beam?. If yes, then replace it because we don't want to have duplicate values. **Case 2**
        - Is the cell in the map and has not been marked as visited by the current beam. **Case 3**

The **Case 3** is what we are interested in. So for example a laser a new laser scan is ready for processing (Say that this is the second, the prior one discovered some cells), since this is a new laser, we clean the visited cells grid and we start processing beam by beam and cell by cell. The first evaluation is the verification to check if that cell is present in the cell probabilities map (Is the map containing the probabilities for each cell provided by each individual laser beam). If the cell is already in the probabilities map but has not been visited for the current beam so we enter this case, because we are appending a new probability to that cell.

Reply:
@kmilo7204 I see. We do visiting checks per scan but we keep the probability grid for all scans.